src/schema/mod.rs

```diff
@@ -0,0 +1,339 @@
+    //
+    // Mapping extension
+    //
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    mapping: Option<mapping::Mapping>,
+    //
+    // Any instance type validation keywords
+    //
+    #[serde(default, rename = "type", deserialize_with = "deserialize_from_str")]
+    type_: Type,
+    #[serde(default, rename = "const", skip_serializing_if = "Option::is_none")]
+    const_: Option<Value>,
+    #[serde(default, rename = "default", skip_serializing_if = "Option::is_none")]
+    default_: Option<Value>,
+    #[serde(default, rename = "enum", skip_serializing_if = "Vec::is_empty")]
+    enum_: Vec<EnumEntry>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    formula: Option<String>,
+    //
+    // Object validation keywords
+    //
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    properties: Vec<Property>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    keys: Option<Box<Schema>>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    values: Option<Box<Schema>>,
+    #[serde(default, rename = "additionalProperties")]
+    additional_properties: bool,
+    //
+    // StringList keywords
+    //
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    separator: Option<String>,
+    //
+    // Annotation keywords
+    //
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    title: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    help: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    warning: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    description: Option<String>,
+    //
+    // Array validation keywords
+    //
+    #[serde(
+        default,
+        skip_serializing_if = "Vec::is_empty",
+        deserialize_with = "deserialize_struct_or_vec"
+    )]
+    items: Vec<Schema>,
+    #[serde(default, rename = "maxItems", skip_serializing_if = "Option::is_none")]
+    max_items: Option<usize>,
+    #[serde(default, rename = "minItems", skip_serializing_if = "Option::is_none")]
+    min_items: Option<usize>,
+    #[serde(default, rename = "uniqueItems")]
+    unique_items: UniqueItems,
+    //
+    // Number validation keywords
+    //
+    #[serde(default, rename = "multipleOf", skip_serializing_if = "Option::is_none")]
+    multiple_of: Option<Number>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    max: Option<Number>,
+    #[serde(default, rename = "exclusiveMax", skip_serializing_if = "Option::is_none")]
+    exclusive_max: Option<Number>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    min: Option<Number>,
+    #[serde(default, rename = "exclusiveMin", skip_serializing_if = "Option::is_none")]
+    exclusive_min: Option<Number>,
+    //
+    // String validation keywords
+    //
+    #[serde(default, rename = "maxLength", skip_serializing_if = "Option::is_none")]
```

Comment:
These sections (like `String validation`) could be split out into separate structs and then included here using `serde(flatten)` - this could allow for an easier time scanning this struct visually ?

Reply:
This one I'm going to close as resolved. Two reasons:

* subjective, I don't like it, because it hides those fields into another structure, which isn't necessary
* flattening is useful in cases where you have frequently grouped keys or if you'd like to capture additional fields
* it says `String validation keywords` in the comment, but it should be `String based types validation keywords` (will push the change), because it's not about `string` only, but these can be used for `hostname`, `date`, `ipv4`, ...